include/fastad_bits/mat.hpp

```diff
@@ -1,16 +1,17 @@
 #pragma once
 #include <iostream>
+#include <iomanip>
 #include <vector>
 #include <algorithm>
 
+#define DEFAULT_PRINT_FIELD_WIDTH 8
+
 namespace ad {
 
 template <class T>
 class Mat;
-
 template <class T>
 std::ostream& operator<<(std::ostream& os, const Mat<T>& mat);
-
 template <class T>
 bool operator==(const Mat<T>& mat1, const Mat<T>& mat2);
 
```

Comment:
```cpp
namespace details {

template <class Matrix>
inline void print(const Matrix& mat, const std::string& header, unsigned int field_width)
{
// move print_at_width definition here and modify accordingly to use mat
}

template <class Matrix>
inline void print(const Matrix& mat, const std::string& header
                                          , unsigned int field_width, unsigned int precision)
{
// move first 2 lines of print_at_precision definition to here
print_at_width(mat, header, field_width + precision);
}

template <class Matrix, class = void>
struct PrintPolicy
{
void print(const std::string& header
               , unsigned int field_width=DEFAULT_PRINT_FIELD_WIDTH) const
{
details::print(static_cast<const Matrix&>(*this), header, field_width);
}
};

template <class Matrix>
struct PrintPolicy<Matrix, std::enable_if_t<std::is_floating_point_v<typename Matrix::value_type>>>
{ 
void print(const std::string& header
                , unsigned int field_width=DEFAULT_PRINT_FIELD_WIDTH
                , unsigned int precision=DEFAULT_PRECISION) const
{
details::print(static_cast<const Matrix&>(*this), header, field_width, precision);
}
};

} // namespace details

template <class T, class PrintPolicy=details::PrintPolicy<Mat<T>>>
class Mat : PrintPolicy
{
public:
using value_type = T;
using PrintPolicy::print;

// continue definition...
```

Reply:
This is really dense template meta-programming xD but essentially what this will allow us to do is inside `Mat`, we don't write any print function. Depending on what `T` is, I want to choose "correct" print function associated with it. I delegate this to `PrintPolicy` to choose for me, which is why I have the primary definition and 1 specialization. I then inherit the "correct" printing policy given `T`. The `print` function in `PrintPolicy` will be publicly exposed inside `Mat`, so user will be able to call it. If `T` is floating point, then user will (only) be able to call the following one:
```cpp
void print(const std::string& header
                , unsigned int field_width=DEFAULT_PRINT_FIELD_WIDTH
                , unsigned int precision=DEFAULT_PRECISION) const
```
Otherwise, they'll (only) be able to call:
```cpp
void print(const std::string& header
               , unsigned int field_width=DEFAULT_PRINT_FIELD_WIDTH) const
```
This kind of design pattern is called Policy-Based Design. It's quite advanced :)